### PR TITLE
Ported the repo to rel-2.3.0

### PR DESCRIPTION
**Pull request by srinjoym-cerebras**
In rel-2.3.0, a new trainer flow and YAML structure was introduced. The corresponding changes were made to the repo.

### CODE REVIEW

File: configs/111m.yaml [hunk 1]
```diff
@@ -16,26 +16,65 @@
...
-learning_rate: 6.0e-4
-warmup_steps: 1525
-backend: "CSX"
-precision_opt_level: 1
-mount_dirs:
-  - /path/to/code/location
-  - /path/to/data/location
-python_paths:
-  - /path/to/code/location
+trainer:
```
Comment (gokulr-cerebras) [inline configs/111m.yaml:19]: Thanks for this, I understand we were able to run on CPU with these changes, but we will need to also verify if we are able to run on CSX. @abhis-cerebras , can you please help here? These are changes made by @srinjoym-cerebras to port our giga gpt model to the new trainer flow. I think he was able to run it on CPU but we will need to verify these changes on CSX.

File: README.md [hunk 2]
```diff
@@ -55,21 +55,21 @@ To train a small 111M parameter model, run the following command. This will work
 (although CPU will be too slow to get that far).
 
 ```bash
-python train.py configs/111m.yaml
+python train.py CSX --mode train --params configs/111b.yaml
```
[the remark this answers is not in the document]
Comment (srinjoym-cerebras) [reply README.md]: Made CSX the default device and updated the README.md

File: model.py [hunk 3]
```diff
@@ -154,15 +164,13 @@ def forward(self, input_ids, labels=None):
...
-                (sequence_length, sequence_length),
-                dtype=x.dtype,
-                device=x.device,
-            ),
-            diagonal=1,
-        ).unsqueeze(0).unsqueeze(0)
-        causal_attention_mask *= torch.finfo(causal_attention_mask.dtype).min
+        causal_attention_mask = create_broadcasted_autoregressive_mask(
+            batch_size=batch_size,
+            num_heads=1,
```
Comment (deepak-cb) [inline model.py:169]:
```suggestion
            num_heads=self.config.heads,
```

File: train.py [hunk 4]
```diff
@@ -11,160 +11,16 @@
...
-
-    for step, batch in enumerate(executor, start=global_step + 1):
-        if step > config.num_steps:
-            break
-        loss = training_step(batch)
-        log_loss(loss, step)
-        save_checkpoint(step)
-
-    logger.info("Training completed successfully!")
+from cerebras.modelzoo.common.utils.run.cli_pytorch import get_params_from_args
```
Comment (deepak-cb) [inline train.py:17]: I don't disagree with this change for getting it to work but I don't think we can continue to claim ~600 lines when we make library calls which hides all the code.

File: train.py [hunk 5]
```diff
@@ -11,160 +11,16 @@
...
 
+def main():
+    params = get_params_from_args()
+    
+    from cerebras.modelzoo.common.run_utils import main
```
Comment (mohitk-cerebras) [inline train.py:22]: do we need to add Cerebras modelzoo dependency in requirements.txt? Please note that this will be a new dependency.

File: model.py (hunk 3 above)
Comment (deepak-cb) [reply model.py:169]: Creating mask explicitly in https://github.com/Cerebras/gigaGPT/pull/5 would be preferable.

File: train.py (hunk 4 above)
Comment (srinjoym-cerebras) [reply train.py:17]: Sorry, could you clarify on what you mean by hiding the code?

Comment (deepak-cb) [reply train.py:17]: Sorry for the late reply. The motivation for maintaining gigaGPT is to demonstrate easy model size scaling on Cerebras Hardware using only simple and readable pytorch code. 

If we delegate all the train.py code to a helper function, we can no longer claim that. One could argue that we scale only by hiding the complexity in a helper function as it is immediately not visible what happens behind `cerebras.modelzoo.common.run_utils.main`.

Comment (srinjoym-cerebras) [reply train.py:17]: ok, got it.
